examples/sktime/flavor.py

```diff
@@ -0,0 +1,639 @@
+"""The ``flavor`` module provides an example for a custom model flavor for ``sktime`` library.
+
+This module exports ``sktime`` models in the following formats:
+
+sktime (native) format
+    This is the main flavor that can be loaded back into sktime, which relies on pickle
+    internally to serialize a model.
+mlflow.pyfunc
+    Produced for use by generic pyfunc-based deployment tools and batch inference.
+
+    The `pyfunc` flavor of the model supports sktime predict methods `predict`,
+    `predict_interval`, `predict_proba`, `predict_quantiles`, `predict_var`.
+
+    The interface for utilizing a sktime model loaded as a `pyfunc` type for
+    generating forecasts requires passing an exogenous regressor as Pandas
+    DataFrame to the `pyfunc.predict()` method (an empty DataFrame must be
+    passed if no exogenous regressor is used). The configuration of predict
+    methods and parameter values passed to the predict methods is defined by
+    a dictionary to be saved as an attribute of the fitted sktime model
+    instance. If no prediction configuration is defined `pyfunc.predict()`
+    will return output from sktime `predict` method. Note that for `pyfunc`
+    flavor the forecasting horizon `fh` must be passed to the fit method.
+
+    Predict methods and parameter values for `pyfunc` flavor can be defined
```

Comment:
This interface seems really complex. Is there a reason why each of the available inference APIs can't be exposed as part of the wrapper instance?

Reply:
Short answer: 

This interface design was favored by sktime PR reviewer. I do not see a strong reason against exposing the inference APIs as part of the wrapper instance, this was actually my original proposal in sktime PR.

For the purpose of this PR and the custom flavor walk-trough would you recommend to simplify the interface?  Happy to get your guidance here   

Long answer:

My original proposal was to design the wrapper class similar to pmdarima wrapper in mlfow, i.e. pass a dataframe with predict method, exogenous regressor and optional parameters. Preferred approach of sktime reviewers was to pass a dictionary which I believe is currently not possible (related issue https://github.com/mlflow/mlflow/issues/2360). Given this limitation they favored the current implementation of passing only the exogenous regressor to the predict method of the wrapper class and assign the prediction configuration (predict method and parameters) as an attribute to the fitted model instance. Also they wanted to allow providing multiple predict methods (e.g. 'predict', 'predict_inteval', 'predict_var') as part of the prediction configuration so that calling wrapper instance predict method will return concatenated prediction output of multiple methods - this is also what makes the interface more complex as you noted.

Reply:
Personally, I like your initial design better :) Can we revisit that? As you mentioned with your thorough research, we do something similar with pmdarima. Using a DataFrame as a currying mechanism and unifying the approach with other flavors is a bit more straightforward. 
This leaves the question as to exogeneous regressor elements. Can these regression terms be stored as an np.ndarray type in a column to allow for passing them (and then subsequently unpacked if the column exists, the row has data of type np.ndarray, and the datatypes of the array are numeric)?

Reply:
Sounds good, I will work on the revised design before addressing your other comments.

On the exogenous regressor, yes we can use the np.ndarray tpye in sktime - according to their documentation (https://github.com/sktime/sktime/blob/main/examples/AA_datatypes_and_datasets.ipynb): _"...sktime allows the user to specify time series as pandas.DataFrame, as pandas.Series, or as a numpy.ndarray"_

I expect the revised design will result in something similar to the below workflow (example for "predict_interval" with exogenous regressor):

```python
import sktime_flavor
from sktime.datatypes import convert
from sktime.forecasting.model_selection import temporal_train_test_split
from sktime.forecasting.naive import NaiveForecaster

model_path = "model"

y, X = load_longley()
y_train, y_test, X_train, X_test = temporal_train_test_split(y, X)

model = NaiveForecaster()
model.fit(y_train, X_train)

sktime_flavor.save_model(
    sktime_model=model, path=model_path
)

X_test_array = convert(X_test, "pd.DataFrame", "np.ndarray")
predict_conf = pd.DataFrame(
    [
        {
            "fh": [1, 2, 3],
            "predict_method": "predict_interval",
            "coverage": [0.1, 0.5, 0.9],
            "X": X_test_array
        }
    ]
)

loaded_pyfunc = sktime_flavor.pyfunc.load_model(model_uri=model_path)
print(loaded_pyfunc.predict(predict_conf))
```

Reply:
Glad to hear this is going in a good direction! Will push some updates probably in the course of next week

Reply:
Implemented as discussed above

Couple of things I noticed:

I omitted sktime "predict_proba" from the example as it would require installation of heavy deep learning dependencies tensorflow and tensorflow-probability  via `pip install sktime[dl]`, wasn't able to install this in my github  codespace and thought it might also get a bit too much for other users who want to follow the custom flavor example. Do you think it is acceptable to leave it out or should we include it?

I realized that using exogenous regressor as numpy array type leads to a JSON serialization error when scoring the model via local REST API. The only way I could make this work is by passing a list and converting this list back to ndarray type in wrapper class. I am not sure if/how this is addressed in other flavors like for example pmdarima, so would be happy to get your guidance here.